src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,286 @@
+/-
+Copyright (c) 2022 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck, David Loeffler
+-/
+import algebra.module.submodule.basic
+import analysis.complex.upper_half_plane.basic
+import linear_algebra.general_linear_group
+import linear_algebra.special_linear_group
+import algebra.direct_sum.ring
+import number_theory.modular
+import geometry.manifold.mfderiv
+import analysis.complex.upper_half_plane.functions_bounded_at_infty
+import number_theory.modular_forms.slash_actions
+
+/-!
+# Modular forms
+
+This file defines modular forms and proves some basic properties about them.
+
+We begin by defining the notion of weakly modular form from which we then we construct the vector
+space of modular forms, cusp forms and prove that the product of two modular forms is a modular
+form (of higher weight).
+-/
+
+universes u v
+
+open complex upper_half_plane
+
+open_locale topological_space manifold upper_half_plane
+
+noncomputable theory
+
+local notation `ℍ'` := (⟨_,
+ upper_half_plane.upper_half_plane_is_open⟩: topological_space.opens ℂ)
+
+local prefix `↑ₘ`:1024 := @coe _ (matrix (fin 2) (fin 2) _) _
+
+local notation `GL(` n `, ` R `)`⁺ := matrix.GL_pos (fin n) R
+
+local notation `SL(` n `, ` R `)` := matrix.special_linear_group (fin n) R
+
+variables (Γ : subgroup SL(2,ℤ)) (k : ℤ)
+
+namespace modular_forms
+
+local notation f `∣[`:73 k:0, A `]`  :72 := slash_action.map ℂ k A f
+
+lemma slash_action_eq_slash (k : ℤ) (A : Γ) (f : ℍ → ℂ) : f ∣[k, A] = slash k A f := rfl
+
+lemma slash_action_eq_slash' (k : ℤ) (A : SL(2, ℤ)) (f : ℍ → ℂ) : f ∣[k, A] = slash k A f := rfl
+
+/-- The space of functions that are weakly modular. -/
+def weakly_modular_submodule (k : ℤ) (Γ : subgroup SL(2,ℤ)) : submodule ℂ (ℍ → ℂ) :=
```

Comment:
The naming convention would be just `weakly_modular`

Reply:
probably `weakly_modular_form` to be consistent with `modular_form` and `cusp_form`